### ISSUE TITLE
[Feature] FPS limit

### ISSUE THREAD

**Issue by Source61** [alievk/avatarify-python#477]
Would be nice to be able to limit FPS to not burn out the GPU.
Planning on using this for up to 12 hours straight streaming for likely multiple sessions, I don't have a stable connection for a cloud notebook, not to mention notebooks will/may stop execution and disconnect once in a while.
Only really need 20-30 FPS, getting 50 on my GTX 3080.

**Comment by JohanAR** [comment]
This code isn't actively maintained right now, so if you have a little bit of Python knowledge it might be easiest if you implement it yourself. The main loop is in afy/cam_fomm.py, and I think it measures frame time already to calculate fps. You ought to be able to add a check to see if a frame limit is set, and then sleep for the remaining time if the frame finished faster than what the limit allows.

Now some unrelated questions: Did you have to do something special to get avatarify-python running, i.e. something not mentioned in the docs? It seemed like the v4l2 module wasn't working for me after I reinstalled with a different distribution, so I haven't done anything with avatarify in quite a while.

Also if you're planning on running this for a long time, have you looked into power limiting you GPU? I'm setting mine to max 200W + some overclocking to still get decent performance in games. I put the script I wrote [in a gist](https://gist.github.com/JohanAR/88a295684116c7ad31f986687f1b282f) if you want to have a look, but you might need to fine tune max clock freq and the offset for you card.

**Comment by Source61** [comment]
> 

Yeah, I've done this before and I've done the same simplified setup that I've done in the past now with avatarify, but the setup of avatarify with TicToc and a dict and a list of fps is different from my setup just using a simple single time lapse and sleeping (1 / max_fps) - lapse, so I'm not sure how to adapt my setup to the show FPS feature (even at capped 20 FPS, which is set in config.yaml, show FPS will still show 48). I prefer simple solutions.

No, I didn't have to do anything extra personally, just followed the docs.
Sometimes I get an issue with the cam, not sure why, but rebooting the system + optionally replugging camera (before reboot) seem to fix the issue it for me.

I have not, I'll take a look thank you. I did notice power consumption and temperature definitely reduced by limiting framerate, from 228W and 65C+ to 125W and 55C on 20 FPS.

**Comment by Source61** [comment]
Just wanted to add that I added a PR shortly after in case you didn't notice it, I'm not sure what github informs about and what not, with the code for optional FPS parsing.
The code adds the config optional fps_limit in config with value 0 by default, n 0 means no limit, n >= 1 limit activated to n fps.
But maybe by not maintained you mean not accepting PRs either.

**Comment by JohanAR** [comment]
Yea I saw it, and I've been meaning to go through it but currently I can't even run avatarify-python (I installed Fedora instead of Xubuntu and now the virtual camera doesn't get created)

**Comment by geekedpriest** [comment]
Any one with basic knowledge on how to run avatarify ? It seems like it stopped working on windows

**Comment by Source61** [comment]
> Yea I saw it, and I've been meaning to go through it but currently I can't even run avatarify-python (I installed Fedora instead of Xubuntu and now the virtual camera doesn't get created)

Gotcha. I run Debian 12 (testing/bookworm) for refr in case that helps.
Is your webcam registered/can you see your webcam through something like VLC?
Thanks for the update.

**Comment by Source61** [comment]
> Now some unrelated questions: Did you have to do something special to get avatarify-python running, i.e. something not mentioned in the docs? It seemed like the v4l2 module wasn't working for me after I reinstalled with a different distribution, so I haven't done anything with avatarify in quite a while.

Just to respond to this again, I just had the two errors I've been having sometimes one after the other, so I'll mention what they say and what I do to "fix" them:
1) modprobe: FATAL: Module v4l2loopback not found in directory /lib/modules/6.0.0-5-amd64
...
RuntimeError: Cannot open camera
2) modprobe: FATAL: Module v4l2loopback not found in directory /lib/modules/6.0.0-5-amd64
...
--- Make sure the v4l2loopback kernel module is loaded ---
sudo modprobe v4l2loopback devices=1
...
FileNotFoundError: device does not exist: /dev/video9
FATAL: exception not rethrown
./run.sh: line 151: 1945 Aborted                     python afy/cam_fomm.py --config $FOMM_CONFIG --checkpoint $FOMM_CKPT --virt-cam $CAMID_VIRT --relative --adapt_scale $@

With both of these errors all I do is just reboot.
In this case I got the first error first, rebooted, then got the second error, rebooted, and now it's working again.
No idea what's happening, but just rebooting once or twice usually does it for me.

Also with the second error I tried running the modprobe command but think I got the first line as an error again, like v4l2loopback is not available/could not be found/is not a module, so I just went ahead and rebooted again, lol.

**Comment by Donvickins** [comment]
Hey @Source61 

Could you make avatarify work with RTX30xx series. 

I have successfully installed and used it on previous laptops but had a Ryzen 7, RTX 3060, it was able to open but as soon as I press X to calibrate, it starts working but pretty laggy.

On task manager, it only using 10% of my GPU.

Note: At first it couldn't detect the GPU, I had to do a few updates to torch and cuda versions, before it was able to use 10% of my GPU.

**Comment by Source61** [comment]
> Hey @Source61
> 
> Could you make avatarify work with RTX30xx series.
> 
> I have successfully installed and used it on previous laptops but had a Ryzen 7 processor, 16gbram ddr4, RTX 3060, hp omen 15 laptop. It was able to open but as soon as I press X to calibrate, it starts working but pretty laggy.
> 
> On task manager, it only using 10% of my GPU.
> 
> Note: At first it couldn't detect the GPU, I had to do a few updates to torch and cuda versions, before it was able to use 10% of my GPU.
> 
> This is on windows 11. Some people here have also had similar issue, with RTX30xx series.

I have a 3080 and it works fine for me on Debian 12 (although sometimes it doesn't work and I have to reboot to make it work), but I always get good FPS performance.
I haven't tested on Windows.
Best of luck.
